source/docs/software/wpilib-tools/robot-simulation/device-sim.rst

```diff
@@ -0,0 +1,117 @@
+      Encoder encoder = new Encoder(2, 3);
+      // create a sim controller for the encoder
+      EncoderSim simEncoder = new EncoderSim(encoder);
+
+   .. code-tab:: cpp
+
+       // create a real encoder object on DIO 2,3
+       Encoder encoder{2, 3};
+       // create a sim controller for the encoder
+       EncoderSim simEncoder{encoder};
+
+Reading and Writing Device Data
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Each simulation class has getter (``getXxx()``/``GetXxx()``) and setter (``setXxx(value)``/``SetXxx(value)``) functions for each field ``Xxx``. The getter functions will return the same as the getter of the regular device class.
+
+.. tabs::
+   .. code-tab:: java
+
+      simEncoder.setCount(100);
+      encoder.getCount(); // 100
+      simEncoder.getCount(); // 100
+
+   .. code-tab:: cpp
+
+      simEncoder.SetCount(100);
+      encoder.GetCount(); // 100
+      simEncoder.GetCount(); // 100
+
+Registering Callbacks
+^^^^^^^^^^^^^^^^^^^^^
+
+In addition to the getters and setters, each field also has a ``registerXxxCallback()`` function that registers a callback to be run whenever the field value changes. These functions return a ``CallbackStore`` object, the callback can be canceled by calling ``close()`` in Java or by automatic destruction in C++. The callbacks accept a string parameter of the name of the field and a ``HALValue`` object containing the value. The value can be retrieved with ``getInt()`` etc.
+
+.. warning:: In C++, do **not** invoke the destructor explicitly! It is invoked automatically on scope exit and causes undefined behavior if explicitly destroyed earlier.
+
+.. warning:: The ``HALValue.getXxx()`` methods are **not** typesafe! For example, calling ``getInt()`` on a ``HALValue`` containing a ``double`` will produce unpredictable results.
+
+.. important:: Make sure to keep a reference to the ``CallbackStore`` object to prevent it being garbage-collected, which will cancel the callback.
+
+.. tabs::
+   .. code-tab:: java
+
+      NotifyCallback callback = (String name, HALValue value) -> {
+         System.out.println("Value of " + name + " is " + value.getInt());
+      }
+      CallbackStore store = simEncoder.registerCountCallback(callback);
+
+      store.close(); // cancel the callback
+
+   .. code-tab:: cpp
+
+      NotifyCallback callback = (String name, HALValue value) {
+         wpi::outs() << "Value of " << name << " is " << value.GetInt() << "\n";
+      }
+      CallbackStore store = simEncoder.RegisterCountCallback(callback);
+      // the callback will be canceled when ``store`` goes out of scope
+
+Simulating Other Devices - The SimDeviceSim Class
+-------------------------------------------------
+
+.. note:: Vendors might implement their connection to the SimDevice API slightly different than described here. They might also provide a simulation class specific for their device class. See your vendor's documentation for more information as to what they support and how.
```

Comment:
Too many admonitions, the idea here is to use them when necessary and avoid having the user overlook them (which they will do if there is too many).